File: backend/src/main/java/com/autocare/service/WorkshopService.java
package com.autocare.service;

import com.autocare.dto.UserProfileResponse;
import com.autocare.dto.WorkshopStatsResponse;
import com.autocare.entity.Role;
import com.autocare.entity.User;
import com.autocare.entity.Workshop;
import com.autocare.exception.ForbiddenException;
import com.autocare.exception.ResourceNotFoundException;
import com.autocare.repository.*;
import com.autocare.security.UserPrincipal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

@Service
@RequiredArgsConstructor
public class WorkshopService {

    private final WorkshopRepository workshopRepository;
    private final UserRepository userRepository;
    private final VehicleRepository vehicleRepository;
    private final AppointmentRepository appointmentRepository;
    private final InvoiceRepository invoiceRepository;

    @Transactional(readOnly = true)
    public WorkshopStatsResponse getWorkshopStats(UserPrincipal principal) {
        if (principal.getRole() != Role.ADMIN) {
            throw new ForbiddenException("Only workshop administrators can access full tenant telemetry.");
        }

        Workshop workshop = workshopRepository.findById(principal.getWorkshopId())
            .orElseThrow(() -> new ResourceNotFoundException("Workshop not found"));

        Integer workshopId = workshop.getWorkshopId();

        long customerCount = userRepository.countByWorkshopIdAndRole(workshopId, Role.CUSTOMER);
        long vehicleCount = vehicleRepository.countByWorkshopId(workshopId);
        long mechanicCount = userRepository.countByWorkshopIdAndRole(workshopId, Role.MECHANIC);
        long scheduledAppointments = appointmentRepository.countByWorkshopIdAndStatus(workshopId, "SCHEDULED")
            + appointmentRepository.countByWorkshopIdAndStatus(workshopId, "IN_PROGRESS");
        long completedAppointments = appointmentRepository.countByWorkshopIdAndStatus(workshopId, "COMPLETED");
        BigDecimal totalRevenue = invoiceRepository.sumTotalRevenueByWorkshopId(workshopId);

        return WorkshopStatsResponse.builder()
            .workshopId(workshopId)
            .workshopName(workshop.getName())
            .workshopAddress(workshop.getAddress())
            .accessCode(workshop.getAccessCode())
            .customerCount(customerCount)
            .vehicleCount(vehicleCount)
            .mechanicCount(mechanicCount)
            .scheduledAppointmentsCount(scheduledAppointments)
            .completedAppointmentsCount(completedAppointments)
            .totalRevenue(totalRevenue != null ? totalRevenue : BigDecimal.ZERO)
            .build();
    }

    @Transactional(readOnly = true)
    public List<UserProfileResponse> getWorkshopMechanics(UserPrincipal principal) {
        List<User> mechanics = userRepository.findAllByWorkshopIdAndRole(principal.getWorkshopId(), Role.MECHANIC);
        return mechanics.stream()
            .map(u -> UserProfileResponse.builder()
                .userId(u.getUserId())
                .workshopId(u.getWorkshop().getWorkshopId())
                .workshopName(u.getWorkshop().getName())
                .email(u.getEmail())
                .firstName(u.getFirstName())
                .lastName(u.getLastName())
                .role(u.getRole())
                .build())
            .toList();
    }
}

File: backend/src/services/workshopService.js
/**
 * AutoCare AI - Workshop Administration & Telemetry Service
 * Aggregates multi-tenant statistics and staff rosters using raw SQL
 */

const workshopRepo = require('../repositories/workshopRepo');
const userRepo = require('../repositories/userRepo');
const vehicleRepo = require('../repositories/vehicleRepo');
const appointmentRepo = require('../repositories/appointmentRepo');
const invoiceRepo = require('../repositories/invoiceRepo');
const {
  NotFoundError,
  ForbiddenError
} = require('../middleware/errorHandler');

class WorkshopService {
  /**
   * Aggregate workshop performance metrics (Admin only).
   */
  async getWorkshopStats(userPrincipal) {
    if (userPrincipal.role !== 'ADMIN') {
      throw new ForbiddenError('Only workshop administrators can access full tenant telemetry.');
    }

    const workshop = await workshopRepo.findById(null, userPrincipal.workshopId);
    if (!workshop) {
      throw new NotFoundError('Workshop not found');
    }

    const workshopId = userPrincipal.workshopId;

    const [
      customerCount,
      vehicleCount,
      mechanicCount,
      scheduledCount,
      inProgressCount,
      completedCount,
      totalRevenue
    ] = await Promise.all([
      userRepo.countByWorkshopIdAndRole(null, workshopId, 'CUSTOMER'),
      vehicleRepo.countByWorkshopId(null, workshopId),
      userRepo.countByWorkshopIdAndRole(null, workshopId, 'MECHANIC'),
      appointmentRepo.countByWorkshopIdAndStatus(null, workshopId, 'SCHEDULED'),
      appointmentRepo.countByWorkshopIdAndStatus(null, workshopId, 'IN_PROGRESS'),
      appointmentRepo.countByWorkshopIdAndStatus(null, workshopId, 'COMPLETED'),
      invoiceRepo.sumTotalRevenueByWorkshopId(null, workshopId)
    ]);

    return {
      workshopId: workshop.workshopId,
      workshopName: workshop.name,
      workshopAddress: workshop.address || '',
      accessCode: workshop.accessCode,
      customerCount,
      vehicleCount,
      mechanicCount,
      scheduledAppointmentsCount: scheduledCount + inProgressCount,
      completedAppointmentsCount: completedCount,
      totalRevenue
    };
  }

  /**
   * Fetch list of certified mechanics belonging to the workshop.
   */
  async getWorkshopMechanics(userPrincipal) {
    const mechanics = await userRepo.findAllByWorkshopIdAndRole(null, userPrincipal.workshopId, 'MECHANIC');
    return mechanics.map(u => ({
      userId: u.userId,
      workshopId: u.workshopId,
      workshopName: u.workshopName,
      email: u.email,
      firstName: u.firstName,
      lastName: u.lastName,
      role: u.role,
      employeeCode: u.employeeCode || undefined
    }));
  }
}

module.exports = new WorkshopService();
